### PR TITLE
Feat: configurable keypair via settings

### LINKED ISSUE #27: Endpoint to host server's public key
Supporting a request of the server's public key directly may help simplify the benefits app.

### LINKED ISSUE #135: Make keypair configurable
As part of moving this app towards production-ready, we need to be able to configure the keypair used as part of the Eligibility Verification token work.

We currently [hardcode paths to the sample keys](https://github.com/cal-itp/eligibility-server/blob/main/eligibility_server/verify.py#L21):

```
with open("./keys/server.key", "rb") as pemfile:
    server_private_key = jwk.JWK.from_pem(pemfile.read())
with open("./keys/client.pub", "rb") as pemfile:
    client_public_key = jwk.JWK.from_pem(pemfile.read())
```

Let's move these into a new couple of settings variables to allow them being overridden.

### PR DESCRIPTION
**Pull request by thekaveman**
Closes #135 

Introduces settings for the client public key, and the server keypair. All can be loaded from a local file or remote URL, assuming PEM text formatting of the keys. For remote URL, a simple anonymous GET request is made (no auth etc.)

Also introduces an endpoint for the server's _public_ key; Closes #27.

I ended up cleaning up the existing tests for the `Verify` resource class - I was having issues getting them to pass related to the mocked config (and of course wanted to mock my own config to test it!) 

The change I made was minor, to use the [`mocker.patch.dict`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch.dict) helper directly on the app config, which seems to be simpler and keeps the app config as a real object rather than a Mock.

Later, I refactored everything I had done in the `Verify` class into a new module `keypair`, but I left the refactors of the tests since they seemed nice.

### CODE REVIEW

File: tests/test_verify.py [hunk 1]
```diff
@@ -1,30 +1,84 @@
...
-    assert response[1] == 200
+    assert response_code == 200
 
 
-def test_Verify_get_response_sub_format_no_match(mocker, mock_config):
-    mocked_config = mock_config({"SUB_FORMAT_REGEX": r"^[A-Z]\d{7}$", "INPUT_HASH_ALGO": ""})
-    mocker.patch("eligibility_server.verify.app.app.config", new=mocked_config)
+def test_Verify_get_response_sub_format_no_match(mocker):
+    mocked_config = {"SUB_FORMAT_REGEX": r"^[A-Z]\d{7}$", "INPUT_HASH_ALGO": ""}
+    mocker.patch.dict("eligibility_server.verify.app.app.config", mocked_config)
```
Comment (angela-tran) [inline tests/test_verify.py:31]: Cool, this reads much nicer!

File: tests/test_verify.py [hunk 2]
```diff
@@ -1,30 +1,84 @@
...
+    with open(default_path, "rb") as df:
+        tf.write(df.read())
+        tf.seek(0)
+
+    mocked_config = {key_path_setting: tf.name}
+    mocker.patch.dict("eligibility_server.verify.app.app.config", mocked_config)
+
+    verify = Verify()
+    # tempfile has been read, close it
+    tf.close()
```
Comment (angela-tran) [inline tests/test_verify.py:76]: (non-blocking) Is it possible to say `with NamedTemporaryFile("wb") as tf:` on line 62 so we don't have to close the temp file explicitly?

**Review by angela-tran — COMMENTED** [review]
These changes overall look great to me. 

Would it make sense to try to get #128 merged in first? Then, this PR can update `docs/configuration/settings.md` to make the section headers match the comments in `eligibility_server/settings.py` and to add `CLIENT_KEY_PATH` and `SERVER_KEY_PATH`.

**Comment by thekaveman** [comment]
@angela-tran 

> Would it make sense to try to get https://github.com/cal-itp/eligibility-server/pull/128 merged in first?

My thoughts exactly! Definitely want to add these configs to the new docs.

**Comment by thekaveman** [comment]
Rebase incoming, with docs for the new settings.

File: tests/test_verify.py (hunk 2 above)
Comment (thekaveman) [reply tests/test_verify.py]: Done in the rebase.

**Comment by thekaveman** [comment]
I have one more change incoming! I want to solve #27 with this one too 😈

File: eligibility_server/verify.py [hunk 3]
```diff
@@ -4,29 +4,27 @@
...
-with open("./keys/client.pub", "rb") as pemfile:
-    client_public_key = jwk.JWK.from_pem(pemfile.read())
-
 
 class Verify(Resource):
     def __init__(self):
-        """Initialize Verify class with a Database and optionally a Hash"""
+        """Initialize Verify class with a keypair and Database"""
+        self.client_public_key = keypair.get_client_public_key()
+        self.server_private_key = keypair.get_server_private_key()
```
Comment (thekaveman) [inline eligibility_server/verify.py:26]: Store the needed keys as instance attributes

File: keys/server.pub [hunk 4]
```diff
@@ -0,0 +1,9 @@
+-----BEGIN PUBLIC KEY-----
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAyYo6Pe9OSfPGX0oQXyLA
+blOwrMgc/j1JlF07b1ahq1lc3FH0XEk3Dzqbt9NuQs8hz6493vBNtNWTpVmvbGe4
+VX3UjhpEARhN3m4jf/Z2OEuDt2A9q19NLSjgeyhieLkYLwN1ezYXrkn7cfOngcJM
+nGDXp45CaA+g3DzasrjETnKUdqecCzJ3FJ/RRwfibrju7eS/8s6H03nvydzeAJzT
+kEv7Fic2JJEUhh2rJhyLxt+qKkIYeBG+5fBri4miaS8FPnD/yjZzEAFsQc7n0dGq
+DAhSJS8tYNmXFmGlaCfRUBNV3mvOx0vFPuH21WQ5KKZxZP0e64/uQdotbPIImiyR
+JwIDAQAB
+-----END PUBLIC KEY-----
```
Comment (thekaveman) [inline keys/server.pub:9]: Added the public key corresponding to our sample private key. Copied from Benefits.

File: requirements.txt [hunk 5]
```diff
@@ -3,3 +3,4 @@ Flask==2.2.2
 Flask-RESTful==0.3.9
 Flask-SQLAlchemy==2.5.1
 jwcrypto==1.3.1
+requests==2.28.1
```
Comment (thekaveman) [inline requirements.txt:6]: Needed to get remote key URLs

File: tests/conftest.py [hunk 6]
```diff
@@ -11,16 +11,3 @@ def flask():
 @pytest.fixture
 def client():
     return app.test_client()
```
Comment (thekaveman) [inline tests/conftest.py:13]: Modified the tests a bit so this fixture wasn't needed

File: tests/test_keypair.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+KEY_PATH_SETTINGS = ["CLIENT_KEY_PATH", "SERVER_PRIVATE_KEY_PATH", "SERVER_PUBLIC_KEY_PATH"]
+
+REMOTE_KEYS = list(
+    zip(
+        (
+            "https://raw.githubusercontent.com/cal-itp/eligibility-server/main/keys/client.pub",
+            "https://raw.githubusercontent.com/cal-itp/eligibility-server/main/keys/server.key",
+            # TODO: omitted until file exists in main branch
+            # "https://raw.githubusercontent.com/cal-itp/eligibility-server/main/keys/server.pub",
```
Comment (thekaveman) [inline tests/test_keypair.py:30]: We need the public key (new in this PR) to exist on the main branch before this test case will pass.

Hence the funk `zip()` order with URLs first. Once we have the file on `main`, we can uncomment this line and swap back to a more normal `key_path_setting, key_path` order.

File: tests/test_keypair.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+        # check that there was a call to open the tempfile path
+        assert mocker.call(tf.name, "rb") in open_spy.call_args_list
+        assert key
+        assert key.key_id
+        assert key.key_type == "RSA"
+
+
+@pytest.mark.usefixtures("reset_cache")
+@pytest.mark.parametrize("key_path,key_path_setting", REMOTE_KEYS)
```
Comment (thekaveman) [inline tests/test_keypair.py:90]: Here's where we use the above remote keys, with the public key currently commented out (so only 2 test cases run)

**Comment by thekaveman** [comment]
OK @angela-tran @machikoyasuda - I promise I am done with this PR.

Since it is a little bigger, I went through and reviewed, leaving some comments on key areas to look at. LMK if you want to chat about anything!

File: docs/configuration/settings.md [hunk 9]
```diff
@@ -60,22 +79,26 @@ The value used for `alg` in the JOSE header of the [JWS](https://jwcrypto.readth
...
+The path to the server's private key, stored as a PEM text file. Used to decrypt the client's request and sign the server's response.
 
-## Database settings
+Can be a path to a local file or a remote URL. For URLs, a simple anonymous GET request is made.
 
-### `SQLALCHEMY_DATABASE_URI`
+### `SERVER_PUBLIC_KEY_PATH`
 
-The URI that should be used for database connection.
+The path to the public key corresponding to the [server's private key](#serverprivatekeypath), stored as a PEM text file. Used by clients to encrypt
```
[the remark this answers is not in the document]
Comment (angela-tran) [reply docs/configuration/settings.md:90]: I'm seeing that the actual permalink generated by mkdocs has underscores, so I think this should be `#server_private_key_path`. I'm not sure why VS Code autocomplete doesn't include the underscores 😕 

image

image

**Review by angela-tran — COMMENTED** [review]
I finished reading through all these commits, and I love how organized and atomic they are 💯 

I did notice the email address on them isn't your `@compiler.la` one 👀 
(If it's too much hassle to change it, I say don't worry about it. Just thought I'd mention it though.)

The code changes look great. I left some non-blocking comments for things I saw in the docs.

**Comment by thekaveman** [comment]
> I did notice the email address on them isn't your @compiler.la one 👀

Ooof... that was an annoying `rebase` 🙄 Thanks for catching me!

File: docs/configuration/settings.md [hunk 10]
```diff
@@ -22,9 +22,18 @@ Value passed as a keyword argument for `host` in `app.run`
 
 The log level used for the server's logging.
 
-## Server settings
+## Database settings
 
-These settings configure how the server parses, composes, and validates requests and responses according to the [Eligibility API specification](https://docs.calitp.org/eligibility-api/specification/).
+!!! note
+See other configurable settings from [Flask-SQLAlchemy](https://flask-sqlalchemy.palletsprojects.com/en/2.x/config/).
```
[the remark this answers is not in the document]
Comment (thekaveman) [reply docs/configuration/settings.md]: Resolved in rebase

File: docs/configuration/settings.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (thekaveman) [reply docs/configuration/settings.md]: Bah, thanks! Fixed in rebase and removed the new markdownlint rule that I had added.

**Review by angela-tran — APPROVED** [review]

**Review by machikoyasuda — APPROVED** [review]

**Comment by thekaveman** [comment]
@angela-tran @machikoyasuda rebase incoming to resolve merge conflicts.

**Pull request merged**